contracts/ring.md

```diff
@@ -0,0 +1,340 @@
+  unit (possibly with a better name).
+- Aside from the put function (and similar functions for transactions
+  and events), the contract language is purely functional.
+- Functions modifying the state need to be annotated with the stateful keyword.
+
+To make it convenient to update parts of a deeply nested state Ring
+provides special special syntax for map/record updates.  Open
+question: we likely want to make it possible to have immutable state
+(parameters). Keep separate from mutable state or annotate certain
+fields as immutable?
+
+### Types
+Ring has the following types:
+
+| Type    | Description                     | Example
+| ------- | ------------------------------- | -------:
+| uint    | A 256 bit integer               | ```42```
+| int     | A 256 bit 2-complement integer  | ```-1```
+| address | A 256 bit number given as a hex | ```ff00```
+| bool    | A Boolean                       | ```true```
+| string  | An array of bytes               | ```“Foo”```
+| list    | A homogeneous immutable singly linked list. | ```[1, 2, 3]```
+| tuple   | An ordered heterogeneous array   | ```(42, “Foo”, true)```
+| record  | An immutable key value store with fixed key names and typed values | ``` type balance = { owner: address, value: uint } ```
+| map     | An immutable key value store with dynamic mapping of keys of one type to values of one type | ```type accounts = map(string, address)```
+| state   | A record of blockstate key, value pairs  |
+| transactions | An append only list of blockchain transactions |
+| events   | An append only list of blockchain events (or log entries) |
+
+#### Arrays
+Ring might not have arrays… Ring might have Maps.
+
+####  Variants
+Depending on how we write the compiler, if they are cheap to include
+some like to use variant types, while others might not want them.
+
+####  Refs
+ Ring has no refs.
+
+####  Object
+ Ring will not have objects. (only contracts) 
+
+### Pattern matching
+Pattern matching is probably outside the first iteration of Ring, but
+we definitely want it in the final language.
+
+### Builtins
+#### Events
+Ring has events which logs a structured message to the contract log in
+the resulting blockchain transaction.
+
+```
+event(name: string, value: any)
+```
+
+#### Transactions
+Ring can generate blockchain transactions.
+```
+transaction(tx_type: transaction, argument: record )
+```
+
+#### State
+
+A Ring contract has a state. You must declare the state record type in
+the contract. You can then update the state through mutable record
+update.
+
+```
+type state = { mutable field: string
+             , mutable value: uint};
+let state = {field: “”, value : 0};
+let store_value = (f: string, v: uint) => {
+    state.field = f;
+    state.value = v;
+}
+```
+#### Contract primitives
+```
+creator() : () => address
+```
+The address (pubkey) of the entity signing for the contract creation transaction
+```
+caller() : () => address
```

Comment:
If I call contract A, which then calls contract B, what would `caller()` return in contract B? The address of A, I assume?